airbnb_project/listings/harvester_app/harvester/spiders/listings_spider.py

```diff
@@ -0,0 +1,312 @@
+# The URL requires an Airbnb listing ID to be inserted into the placeholder.
+API_CALL = "https://www.airbnb.ca/api/v3/StaysPdpSections/08e3ad2e3d75c9bede923485718ff2e7f6efe2ca1febb5192d78c51e17e8b4ca?operationName=StaysPdpSections&locale=en-CA&currency=CAD&variables=%7B%22id%22%3A%22{}%22%2C%22pdpSectionsRequest%22%3A%7B%22adults%22%3A%221%22%2C%22layouts%22%3A%5B%22SINGLE_COLUMN%22%5D%7D%7D&extensions=%7B%22persistedQuery%22%3A%7B%22version%22%3A1%2C%22sha256Hash%22%3A%2237d7cbb631196506c3990783fe194d81432d0fbf7362c668e547bb6475e71b37%22%7D%7D"
+
+# API key used for authentication when making API calls to Airbnb's services.
+API_KEY = "d306zoyjsyarp7ifhu67rjxn52tv0t20"
+
+# Zoom level parameter used in the map URL templates to define the zoom level of the map.
+ZOOM_LEVEL = 15.4
+
+
+
+def n(s):
+    """
+    Encode the input string `s` into bytes, then encode it in base64, and decode it back to a UTF-8 string.
+
+    :param s: The input string to be encoded.
+    :return: The base64 encoded string.
+    """
+    return base64.b64encode(s.encode('utf-8')).decode('utf-8')
+
+
+from urllib.parse import quote
+
+
+def t(n, t):
+    """
+    Encode the second string `t` using URL encoding, replace spaces with plus signs, and return the combined format.
+
+    :param n: The first part of the string to be combined.
+    :param t: The string to be URL encoded and combined.
+    :return: A formatted string combining `n` and the encoded version of `t`.
+    """
+    encoded_t = quote(t, safe='').replace('%20', '+').replace('(', '%28').replace(')', '%29')
+    return f"{n}:{encoded_t}"
+
+
+class ListingsSpider(scrapy.Spider):
+    """
+    Spider for scraping Airbnb listings.
+    """
+    # Name of the spider, used by Scrapy to identify this spider
+    name = "listings_spider"
+
+    # List of allowed domains for the spider to crawl
+    allowed_domains = ['airbnb.ca']
+
+    # List to store pagination cursors for navigating through result pages
+    next_page_cursors: [str] = None
+
+    # Counter to keep track of the total number of listings scraped
+    total_listings = 0
+
+    # Set to store IDs of listings that have already been scraped to avoid duplicates. This is a
+    # temporary solution until a database is set up, which will then take care of duplicates etc
+    ALREADY_SCRAPED_LISTINGS = set()
+
+    # Instance of AirBnbURLBuilder to generate URLs for different search parameters
+    URL_BUILDER = AirBnbURLBuilder()
+
+    def start_requests(self):
+        """
+       Generates initial requests for scraping Airbnb listings using coordinates from a JSON file.
+
+       This function performs the following steps:
+       1. Reads already scraped Airbnb listing IDs from a CSV file and stores them in a set.
+       2. Reads coordinates from a JSON file and shuffles them.
+       3. Constructs and returns a list of initial Scrapy requests using template URLs and the read coordinates.
```

Comment:
since there are 3 different tasks here, should this function be refactored into smaller functions?